## Week4
### 【問題1】コンペティション内容の確認

コンペティションのOverviewページ読み、「Home Credit Default Risk」について以下の観点について確認してください。


何を学習し、何を予測するのか


 →　各顧客の特徴量（年齢、年収等）から返済できるかどうかを学習し返済有無を予測する。

どのようなファイルを作りKaggleに提出するか

　→テストセットの各SK_ID_CURRについて、TARGET変数の確率を予測する

提出されたものはどういった指標値で評価されるのか

　→予測された確率と観測された目標の間のROC曲線の下の領域で評価
   ( from sklearn.metrics import roc_auc_score )


### 【問題2】学習と検証
データを簡単に分析、前処理し、学習、検証するまでの一連の流れを作成・実行してください。


検証にはこのコンペティションで使用される評価指標を用いるようにしてください。学習に用いる手法は指定しません。

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# データ取得
train_data = pd.read_csv("application_train.csv")
pd.set_option('display.max_columns', 122)
pd.set_option('display.max_rows', 122)
train_data

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0           100002       1         Cash loans           M            N   
1           100003       0         Cash loans           F            N   
2           100004       0    Revolving loans           M            Y   
3           100006       0         Cash loans           F            N   
4           100007       0         Cash loans           M            N   
...            ...     ...                ...         ...          ...   
307506      456251       0         Cash loans           M            N   
307507      456252       0         Cash loans           F            N   
307508      456253       0         Cash loans           F            N   
307509      456254       1         Cash loans           F            N   
307510      456255       0         Cash loans           F            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
0                    Y             0          202500.0    406597.5   
1                    N             0          270000.0   1293502.5   
2                    Y             0           67500.0    135000.0   
3                    Y             0          135000.0    312682.5   
4                    Y             0          121500.0    513000.0   
...                ...           ...               ...         ...   
307506               N             0          157500.0    254700.0   
307507               Y             0           72000.0    269550.0   
307508               Y             0          153000.0    677664.0   
307509               Y             0          171000.0    370107.0   
307510               N             0          157500.0    675000.0   

        AMT_ANNUITY  AMT_GOODS_PRICE NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
0           24700.5         351000.0   Unaccompanied               Working   
1           35698.5        1129500.0          Family         State servant   
2            6750.0         135000.0   Unaccompanied               Working   
3           29686.5         297000.0   Unaccompanied               Working   
4           21865.5         513000.0   Unaccompanied               Working   
...             ...              ...             ...                   ...   
307506      27558.0         225000.0   Unaccompanied               Working   
307507      12001.5         225000.0   Unaccompanied             Pensioner   
307508      29979.0         585000.0   Unaccompanied               Working   
307509      20205.0         319500.0   Unaccompanied  Commercial associate   
307510      49117.5         675000.0   Unaccompanied  Commercial associate   

                  NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
0       Secondary / secondary special  Single / not married   
1                    Higher education               Married   
2       Secondary / secondary special  Single / not married   
3       Secondary / secondary special        Civil marriage   
4       Secondary / secondary special  Single / not married   
...                               ...                   ...   
307506  Secondary / secondary special             Separated   
307507  Secondary / secondary special                 Widow   
307508               Higher education             Separated   
307509  Secondary / secondary special               Married   
307510               Higher education               Married   

        NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  \
0       House / apartment                    0.018801       -9461   
1       House / apartment                    0.003541      -16765   
2       House / apartment                    0.010032      -19046   
3       House / apartment                    0.008019      -19005   
4       House / apartment                    0.028663      -19932   
...                   ...                         ...         ...   
307506       With parents                    0.032561       -9327   
307507  House / apartment       

In [2]:
# TARGETとの相関係数を確認
corr_matrix = train_data.corr()
corr_matrix["TARGET"].nlargest(10)

TARGET                         1.000000
DAYS_BIRTH                     0.078239
REGION_RATING_CLIENT_W_CITY    0.060893
REGION_RATING_CLIENT           0.058899
DAYS_LAST_PHONE_CHANGE         0.055218
DAYS_ID_PUBLISH                0.051457
REG_CITY_NOT_WORK_CITY         0.050994
FLAG_EMP_PHONE                 0.045982
REG_CITY_NOT_LIVE_CITY         0.044395
FLAG_DOCUMENT_3                0.044346
Name: TARGET, dtype: float64

ベースラインとしては、特徴量にターゲットとの相関係数が大きい"DAYS_BIRTH","REGION_RATING_CLIENT_W_CITY","REGION_RATING_CLIENT"の3つとし、k近傍法で学習・評価をすることとした。

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import roc_auc_score 

# 特徴量とラベルを抽出
X = train_data[["DAYS_BIRTH","REGION_RATING_CLIENT_W_CITY","REGION_RATING_CLIENT"]]
y = train_data["TARGET"]

# データをホールドアウト法で分割
X_train,X_test,y_train,y_test = train_test_split(X,y,train_size=0.75,stratify=y,random_state=1)

# 標準化
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scalered = scaler.transform(X_train)
X_test_scalered = scaler.transform(X_test)

# k近傍法でモデルの学習
model = KNeighborsClassifier(n_neighbors=100)
model.fit(X_train_scalered,y_train)
predicted_label = model.predict_proba(X_test_scalered)
predicted_label = predicted_label[:,1]

# 評価
roc_auc_score(y_test,predicted_label)

0.5744054250139229

### 【問題3】テストデータに対する推定
テストデータ（application_test.csv）に対して推定を行い、Kaggleに提出を行ってください。


正しく提出が行えていれば、精度は低くても構いません。

In [4]:
test_data = pd.read_csv("application_test.csv")
test_data 

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0          100001         Cash loans           F            N               Y   
1          100005         Cash loans           M            N               Y   
2          100013         Cash loans           M            Y               Y   
3          100028         Cash loans           F            N               Y   
4          100038         Cash loans           M            Y               N   
...           ...                ...         ...          ...             ...   
48739      456221         Cash loans           F            N               Y   
48740      456222         Cash loans           F            N               N   
48741      456223         Cash loans           F            Y               Y   
48742      456224         Cash loans           M            N               N   
48743      456250         Cash loans           F            Y               N   

       CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                 0          135000.0    568800.0      20560.5   
1                 0           99000.0    222768.0      17370.0   
2                 0          202500.0    663264.0      69777.0   
3                 2          315000.0   1575000.0      49018.5   
4                 1          180000.0    625500.0      32067.0   
...             ...               ...         ...          ...   
48739             0          121500.0    412560.0      17473.5   
48740             2          157500.0    622413.0      31909.5   
48741             1          202500.0    315000.0      33205.5   
48742             0          225000.0    450000.0      25128.0   
48743             0          135000.0    312768.0      24709.5   

       AMT_GOODS_PRICE NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
0             450000.0   Unaccompanied               Working   
1             180000.0   Unaccompanied               Working   
2             630000.0             NaN               Working   
3            1575000.0   Unaccompanied               Working   
4             625500.0   Unaccompanied               Working   
...                ...             ...                   ...   
48739         270000.0   Unaccompanied               Working   
48740         495000.0   Unaccompanied  Commercial associate   
48741         315000.0   Unaccompanied  Commercial associate   
48742         450000.0          Family  Commercial associate   
48743         270000.0   Unaccompanied               Working   

                 NAME_EDUCATION_TYPE NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0                   Higher education            Married  House / apartment   
1      Secondary / secondary special            Married  House / apartment   
2                   Higher education            Married  House / apartment   
3      Secondary / secondary special            Married  House / apartment   
4      Secondary / secondary special            Married  House / apartment   
...                              ...                ...                ...   
48739  Secondary / secondary special              Widow  House / apartment   
48740  Secondary / secondary special            Married  House / apartment   
48741  Secondary / secondary special            Married  House / apartment   
48742               Higher education            Married  House / apartment   
48743  Secondary / secondary special            Married  House / apartment   

       REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  \
0                        0.018850      -19241          -2329   
1                        0.035792      -18064          -4469   
2                        0.019101      -20038          -4458   
3                        0.026392      -13976          -1866   
4                        0.010032      -13040          -2191   
...                           ...         ...            ...   
48739                    0.002042      -19970          -5169   
48740                  

In [5]:
# 特徴量を抽出
X = test_data[["DAYS_BIRTH","REGION_RATING_CLIENT_W_CITY","REGION_RATING_CLIENT"]]

# 標準化
X_scalered = scaler.transform(X)

# 学習済モデルで予測
predicted_label = model.predict_proba(X_scalered)
predicted_label = predicted_label[:,1].reshape(-1,1)

# 提出できる形にIDとラベルの予測値（確率）を結合する
ID = np.array(test_data["SK_ID_CURR"]).reshape(-1,1)
submission1_values = np.concatenate([ID,predicted_label],axis=1)
submission1_values = pd.DataFrame(submission1_values,columns=["SK_ID_CURR","TARGET"])
submission1_values["SK_ID_CURR"] = submission1_values["SK_ID_CURR"].astype("int32")
print(submission1_values.head())

# csvに変換
submission1_values.to_csv('submission1.csv',index=False)

   SK_ID_CURR  TARGET
0      100001    0.09
1      100005    0.05
2      100013    0.04
3      100028    0.07
4      100038    0.13


スコア：0.57607

### 【問題4】特徴量エンジニアリング
精度を上げるために以下のような観点で 特徴量エンジニアリング（Feature Engineering） を行ってください。


どの特徴量を使うか
どう前処理をするか

何をした時に検証データに対する評価指標がどのようになったかをまとめてください。最低5パターンの学習・検証を行ってください。


精度が高かったものに関してはテストデータに対しても推定を行い、Kaggleに提出を行ってください。

初めに特徴量を全て（数値データのみ）を使用し確認する。

In [7]:
only_num = train_data.select_dtypes(include='number')

In [10]:
only_num = only_num.dropna(how='any')

In [11]:
# 特徴量とラベルを抽出
X = only_num.drop(["TARGET"],axis=1)
y = only_num["TARGET"]

# データをホールドアウト法で分割
X_train,X_test,y_train,y_test = train_test_split(X,y,train_size=0.75,stratify=y,random_state=1)

# 標準化
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scalered = scaler.transform(X_train)
X_test_scalered = scaler.transform(X_test)

# k近傍法でモデルの学習
model = KNeighborsClassifier(n_neighbors=100)
model.fit(X_train_scalered,y_train)
predicted_label = model.predict_proba(X_test_scalered)
predicted_label = predicted_label[:,1]

# 評価
roc_auc_score(y_test,predicted_label)

0.6459258382175911

スコアは0.64という結果となった。ここから特徴量を選択してどのように変化するか確認する。
前回で相関係数が正に大きい特徴量は確認したので以下を確認する。

>相関係数が-1に近い特徴量を利用する

In [13]:
corr_matrix["TARGET"].nsmallest(10)

EXT_SOURCE_3                 -0.178919
EXT_SOURCE_2                 -0.160472
EXT_SOURCE_1                 -0.155317
DAYS_EMPLOYED                -0.044932
FLOORSMAX_AVG                -0.044003
FLOORSMAX_MEDI               -0.043768
FLOORSMAX_MODE               -0.043226
AMT_GOODS_PRICE              -0.039645
REGION_POPULATION_RELATIVE   -0.037227
ELEVATORS_AVG                -0.034199
Name: TARGET, dtype: float64

In [14]:
X2.isnull().sum()　　#欠損値の確認

SyntaxError: invalid character in identifier (<ipython-input-14-08e39e2bcbad>, line 1)

In [20]:
X2 = train_data.dropna(how="any")
X2.isnull().sum()

SK_ID_CURR                      0
TARGET                          0
NAME_CONTRACT_TYPE              0
CODE_GENDER                     0
FLAG_OWN_CAR                    0
FLAG_OWN_REALTY                 0
CNT_CHILDREN                    0
AMT_INCOME_TOTAL                0
AMT_CREDIT                      0
AMT_ANNUITY                     0
AMT_GOODS_PRICE                 0
NAME_TYPE_SUITE                 0
NAME_INCOME_TYPE                0
NAME_EDUCATION_TYPE             0
NAME_FAMILY_STATUS              0
NAME_HOUSING_TYPE               0
REGION_POPULATION_RELATIVE      0
DAYS_BIRTH                      0
DAYS_EMPLOYED                   0
DAYS_REGISTRATION               0
DAYS_ID_PUBLISH                 0
OWN_CAR_AGE                     0
FLAG_MOBIL                      0
FLAG_EMP_PHONE                  0
FLAG_WORK_PHONE                 0
FLAG_CONT_MOBILE                0
FLAG_PHONE                      0
FLAG_EMAIL                      0
OCCUPATION_TYPE                 0
CNT_FAM_MEMBER

ここではEXT_SOURCE_3、EXT_SOURCE_2 、EXT_SOURCE_1　を用いて学習を行う。

In [22]:
# 特徴量とラベルを抽出
train_data = train_data.dropna(how='any') # 欠損値があるため削除
X2 = train_data[["EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1"]]
y2 = train_data["TARGET"]

# データをホールドアウト法で分割
X2_train,X2_test,y2_train,y2_test = train_test_split(X2,y2,train_size=0.75,stratify=y2,random_state=1)

# 標準化
scaler = StandardScaler()
scaler.fit(X2_train)
X2_train_scalered = scaler.transform(X2_train)
X2_test_scalered = scaler.transform(X2_test)

# k近傍法でモデルの学習
model_2 = KNeighborsClassifier(n_neighbors=100)
model_2.fit(X2_train_scalered,y2_train)
predicted_label_2 = model_2.predict_proba(X2_test_scalered)
predicted_label_2 = predicted_label_2[:,1]

# 評価
roc_auc_score(y2_test,predicted_label_2)

0.7261714470109715

スコアは前回の0.574よりも上昇した。これは、相関係数の絶対値が大きい特徴量を使用したためめと考えられる。"EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1"

次に、相関係数が正で最も大きい"DAYS_BIRTH"、負で最も小さい（絶対値大きい）"EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1"を組み合わせて確認する。

In [23]:
# 特徴量とラベルを抽出
train_data = train_data.dropna(how='any') # 欠損値があるため削除
X3 = train_data[["DAYS_BIRTH","EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1"]]
y3 = train_data["TARGET"]

# データをホールドアウト法で分割
X3_train,X3_test,y3_train,y3_test = train_test_split(X3,y3,train_size=0.75,stratify=y3,random_state=1)

# 標準化
scaler = StandardScaler()
scaler.fit(X3_train)
X3_train_scalered = scaler.transform(X3_train)
X3_test_scalered = scaler.transform(X3_test)

# k近傍法でモデルの学習
model_3 = KNeighborsClassifier(n_neighbors=100)
model_3.fit(X3_train_scalered,y3_train)
predicted_label_3 = model_3.predict_proba(X3_test_scalered)
predicted_label_3 = predicted_label_3[:,1]

# 評価
roc_auc_score(y3_test,predicted_label_3)

0.7241752592792713

"DAYS_BIRTH""EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1"の組み合わせた特徴量でもある程度高いスコア0.724を確認できた。
既存の特徴量でターゲットとの相関係数が高い特徴量で学習させてるとある程度高いスコアを確認することができた。

最後にこの"EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1"については各特徴量の掛け算や各の平均したものを特徴量として生成した。
また、年齢単位（'DAYS_BIRTH'）や　金属年数単位'DAYS_EMPLOYED'、年収単位で　代表的な特徴量を生成した。こちらを加えて確認をしてみる。

In [27]:
df= pd.read_csv("application_train.csv")

df['EXT_SOURCE_3'] = df['EXT_SOURCE_3'].fillna(df['EXT_SOURCE_3'].mean())
df['EXT_SOURCE_2'] = df['EXT_SOURCE_2'].fillna(df['EXT_SOURCE_2'].mean())
df['EXT_SOURCE_1'] = df['EXT_SOURCE_1'].fillna(df['EXT_SOURCE_1'].mean())
df['NEW_EXT_SOURCES_MEAN'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].mean(axis=1)
df['NEW_SCORES_STD'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].std(axis=1)
df['NEW_SCORES_STD'] = df['NEW_SCORES_STD'].fillna(df['NEW_SCORES_STD'].mean())


df['NEW_CAR_TO_BIRTH_RATIO'] = df['OWN_CAR_AGE'] / df['DAYS_BIRTH']
df['NEW_CAR_TO_BIRTH_RATIO'] = df['NEW_CAR_TO_BIRTH_RATIO'].fillna(df['NEW_CAR_TO_BIRTH_RATIO'].mean())
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['DAYS_LAST_PHONE_CHANGE'] / df['DAYS_BIRTH']
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['NEW_PHONE_TO_BIRTH_RATIO'].fillna(df['NEW_PHONE_TO_BIRTH_RATIO'].mean())

df['NEW_EMPLOY_TO_BIRTH_RATIO'] = df['DAYS_EMPLOYED'] / df['DAYS_BIRTH']

df['NEW_CREDIT_TO_INCOME_RATIO'] = df['AMT_CREDIT'] / df['AMT_INCOME_TOTAL']

In [28]:
train = df[["DAYS_BIRTH","EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1",'NEW_EXT_SOURCES_MEAN','NEW_SCORES_STD','NEW_PHONE_TO_BIRTH_RATIO','NEW_EMPLOY_TO_BIRTH_RATIO','NEW_PHONE_TO_BIRTH_RATIO','NEW_CREDIT_TO_INCOME_RATIO']]

In [29]:
df.isnull().sum()

SK_ID_CURR                    0
TARGET                        0
NAME_CONTRACT_TYPE            0
CODE_GENDER                   0
FLAG_OWN_CAR                  0
                             ..
NEW_SCORES_STD                0
NEW_CAR_TO_BIRTH_RATIO        0
NEW_PHONE_TO_BIRTH_RATIO      0
NEW_EMPLOY_TO_BIRTH_RATIO     0
NEW_CREDIT_TO_INCOME_RATIO    0
Length: 128, dtype: int64

In [30]:
# 特徴量とラベルを抽出
X4 = train
y4 = df["TARGET"]

# データをホールドアウト法で分割
X4_train,X4_test,y4_train,y4_test = train_test_split(X4,y4,train_size=0.75,stratify=y4,random_state=1)

# 標準化
scaler = StandardScaler()
scaler.fit(X4_train)
X4_train_scalered = scaler.transform(X4_train)
X4_test_scalered = scaler.transform(X4_test)

# k近傍法でモデルの学習
model_4 = KNeighborsClassifier(n_neighbors=100)
model_4.fit(X4_train_scalered,y4_train)
predicted_label_4 = model_4.predict_proba(X4_test_scalered)
predicted_label_4 = predicted_label_4[:,1]

# 評価
roc_auc_score(y4_test,predicted_label_4)

0.7153584565656919

スコア結果より、こちらでも同程度のスコアを確認することができた。今回はこの最後の学習モデルでkaggleに提出する。

In [31]:
df = pd.read_csv("application_test.csv")

df['EXT_SOURCE_3'] = df['EXT_SOURCE_3'].fillna(df['EXT_SOURCE_3'].mean())
df['EXT_SOURCE_2'] = df['EXT_SOURCE_2'].fillna(df['EXT_SOURCE_2'].mean())
df['EXT_SOURCE_1'] = df['EXT_SOURCE_1'].fillna(df['EXT_SOURCE_1'].mean())
df['NEW_EXT_SOURCES_MEAN'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].mean(axis=1)
df['NEW_SCORES_STD'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].std(axis=1)
df['NEW_SCORES_STD'] = df['NEW_SCORES_STD'].fillna(df['NEW_SCORES_STD'].mean())


df['NEW_CAR_TO_BIRTH_RATIO'] = df['OWN_CAR_AGE'] / df['DAYS_BIRTH']
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['DAYS_LAST_PHONE_CHANGE'] / df['DAYS_BIRTH']
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['NEW_PHONE_TO_BIRTH_RATIO'].fillna(df['NEW_PHONE_TO_BIRTH_RATIO'].mean())

df['NEW_EMPLOY_TO_BIRTH_RATIO'] = df['DAYS_EMPLOYED'] / df['DAYS_BIRTH']

df['NEW_CREDIT_TO_INCOME_RATIO'] = df['AMT_CREDIT'] / df['AMT_INCOME_TOTAL']

In [32]:
# 特徴量を抽出
X = df[["DAYS_BIRTH","EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1",'NEW_EXT_SOURCES_MEAN','NEW_SCORES_STD','NEW_PHONE_TO_BIRTH_RATIO','NEW_EMPLOY_TO_BIRTH_RATIO','NEW_PHONE_TO_BIRTH_RATIO','NEW_CREDIT_TO_INCOME_RATIO']]

# 標準化
X_scalered = scaler.transform(X)

# 学習済モデルで予測
predicted_label = model_4.predict_proba(X_scalered)
predicted_label = predicted_label[:,1].reshape(-1,1)

# 提出できる形にIDとラベルの予測値（確率）を結合する
ID = np.array(df["SK_ID_CURR"]).reshape(-1,1)
submission2_values = np.concatenate([ID,predicted_label],axis=1)
submission2_values = pd.DataFrame(submission2_values,columns=["SK_ID_CURR","TARGET"])
submission2_values["SK_ID_CURR"] = submission2_values["SK_ID_CURR"].astype("int32")
print(submission2_values.head())

# csvに変換
submission2_values.to_csv('submission2.csv',index=False)

   SK_ID_CURR  TARGET
0      100001    0.02
1      100005    0.19
2      100013    0.02
3      100028    0.06
4      100038    0.18


スコア：0.69537

### ※自主取り組み（参考）

In [34]:
df= pd.read_csv("application_train.csv")

df['EXT_SOURCE_3'] = df['EXT_SOURCE_3'].fillna(df['EXT_SOURCE_3'].mean())
df['EXT_SOURCE_2'] = df['EXT_SOURCE_2'].fillna(df['EXT_SOURCE_2'].mean())
df['EXT_SOURCE_1'] = df['EXT_SOURCE_1'].fillna(df['EXT_SOURCE_1'].mean())
df['NEW_EXT_SOURCES_MEAN'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].mean(axis=1)
df['NEW_SCORES_STD'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].std(axis=1)
df['NEW_SCORES_STD'] = df['NEW_SCORES_STD'].fillna(df['NEW_SCORES_STD'].mean())


df['NEW_CAR_TO_BIRTH_RATIO'] = df['OWN_CAR_AGE'] / df['DAYS_BIRTH']
df['NEW_CAR_TO_BIRTH_RATIO'] = df['NEW_CAR_TO_BIRTH_RATIO'].fillna(df['NEW_CAR_TO_BIRTH_RATIO'].mean())
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['DAYS_LAST_PHONE_CHANGE'] / df['DAYS_BIRTH']
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['NEW_PHONE_TO_BIRTH_RATIO'].fillna(df['NEW_PHONE_TO_BIRTH_RATIO'].mean())

df['NEW_EMPLOY_TO_BIRTH_RATIO'] = df['DAYS_EMPLOYED'] / df['DAYS_BIRTH']

df['NEW_CREDIT_TO_INCOME_RATIO'] = df['AMT_CREDIT'] / df['AMT_INCOME_TOTAL']

In [35]:
import lightgbm as lgb
from sklearn.model_selection import StratifiedKFold

kf = StratifiedKFold(n_splits=4)
score_probs_gbm = []

X = df[["DAYS_BIRTH","EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1",'NEW_EXT_SOURCES_MEAN','NEW_SCORES_STD','NEW_PHONE_TO_BIRTH_RATIO','NEW_EMPLOY_TO_BIRTH_RATIO','NEW_PHONE_TO_BIRTH_RATIO','NEW_CREDIT_TO_INCOME_RATIO']]
y = df["TARGET"]

for tr_idx, va_idx in kf.split(X,y):
    # 学習データとバリデーションに分ける
    tr_X, va_X = X.iloc[tr_idx], X.iloc[va_idx]
    tr_y, va_y = y.iloc[tr_idx], y.iloc[va_idx]
    
    # 標準化
    sscaler = StandardScaler()
    sscaler.fit(tr_X)
    tr_X_scalered = sscaler.transform(tr_X)
    va_X_scalered = sscaler.transform(va_X)
    
    # lightgbm が扱うデータセットの形式に直す
    dtrain = lgb.Dataset(tr_X_scalered, label=tr_y)
    dtest = lgb.Dataset(va_X_scalered, label=va_y, reference= dtrain)
    
    #モデル学習
    params= {'objective':'binary','learning_rate':0.1,'reg_alpha':0,'leg_lambda':1,'min_child_weight':1,'max_depth':5,'subsample':0.8,'colsample':0.8,'colsample_bytree':0.8,'ramdom_state':71}
    gbm = lgb.train(params,
                    dtrain,
                    valid_sets=dtest,
                    num_boost_round=100,
                    verbose_eval=10
                    )
    
    #確率予測を返す
    predicted_label_gbm = gbm.predict(va_X_scalered)
    
    #スコア算出
    score_gbm = roc_auc_score(va_y,predicted_label_gbm)
    
    #スコアを記録
    score_probs_gbm.append(score_gbm)

print(score_probs_gbm)
print(np.mean(score_probs_gbm))

/Users/mishibatoshihiro/.pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/lightgbm/__init__.py:48: UserWarning: Starting from version 2.2.1, the library file in distribution wheels for macOS is built by the Apple Clang (Xcode_8.3.3) compiler.
This means that in case of installing LightGBM from PyPI via the ``pip install lightgbm`` command, you don't need to install the gcc compiler anymore.
Instead of that, you need to install the OpenMP library, which is required for running LightGBM on the system with the Apple Clang compiler.
You can install the OpenMP library by the following command: ``brew install libomp``.
  "You can install the OpenMP library by the following command: ``brew install libomp``.", UserWarning)


[10]	valid_0's binary_logloss: 0.25839
[20]	valid_0's binary_logloss: 0.254631
[30]	valid_0's binary_logloss: 0.253487
[40]	valid_0's binary_logloss: 0.253007
[50]	valid_0's binary_logloss: 0.252704
[60]	valid_0's binary_logloss: 0.252629
[70]	valid_0's binary_logloss: 0.252511
[80]	valid_0's binary_logloss: 0.252481
[90]	valid_0's binary_logloss: 0.252521
[100]	valid_0's binary_logloss: 0.252549
[10]	valid_0's binary_logloss: 0.258437
[20]	valid_0's binary_logloss: 0.254678
[30]	valid_0's binary_logloss: 0.253477
[40]	valid_0's binary_logloss: 0.252967
[50]	valid_0's binary_logloss: 0.252678
[60]	valid_0's binary_logloss: 0.252469
[70]	valid_0's binary_logloss: 0.252343
[80]	valid_0's binary_logloss: 0.252305
[90]	valid_0's binary_logloss: 0.252262
[100]	valid_0's binary_logloss: 0.252196
[10]	valid_0's binary_logloss: 0.258656
[20]	valid_0's binary_logloss: 0.254824
[30]	valid_0's binary_logloss: 0.253725
[40]	valid_0's binary_logloss: 0.253136
[50]	valid_0's binary_logloss: 0.252888

In [37]:
df = pd.read_csv("application_test.csv")

df['EXT_SOURCE_3'] = df['EXT_SOURCE_3'].fillna(df['EXT_SOURCE_3'].mean())
df['EXT_SOURCE_2'] = df['EXT_SOURCE_2'].fillna(df['EXT_SOURCE_2'].mean())
df['EXT_SOURCE_1'] = df['EXT_SOURCE_1'].fillna(df['EXT_SOURCE_1'].mean())
df['NEW_EXT_SOURCES_MEAN'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].mean(axis=1)
df['NEW_SCORES_STD'] = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].std(axis=1)
df['NEW_SCORES_STD'] = df['NEW_SCORES_STD'].fillna(df['NEW_SCORES_STD'].mean())


df['NEW_CAR_TO_BIRTH_RATIO'] = df['OWN_CAR_AGE'] / df['DAYS_BIRTH']
df['NEW_CAR_TO_BIRTH_RATIO'] = df['NEW_CAR_TO_BIRTH_RATIO'].fillna(df['NEW_CAR_TO_BIRTH_RATIO'].mean())
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['DAYS_LAST_PHONE_CHANGE'] / df['DAYS_BIRTH']
df['NEW_PHONE_TO_BIRTH_RATIO'] = df['NEW_PHONE_TO_BIRTH_RATIO'].fillna(df['NEW_PHONE_TO_BIRTH_RATIO'].mean())

df['NEW_EMPLOY_TO_BIRTH_RATIO'] = df['DAYS_EMPLOYED'] / df['DAYS_BIRTH']

df['NEW_CREDIT_TO_INCOME_RATIO'] = df['AMT_CREDIT'] / df['AMT_INCOME_TOTAL']

In [38]:
# 特徴量を抽出
X = df[["DAYS_BIRTH","EXT_SOURCE_3","EXT_SOURCE_2","EXT_SOURCE_1",'NEW_EXT_SOURCES_MEAN','NEW_SCORES_STD','NEW_PHONE_TO_BIRTH_RATIO','NEW_EMPLOY_TO_BIRTH_RATIO','NEW_PHONE_TO_BIRTH_RATIO','NEW_CREDIT_TO_INCOME_RATIO']]

# 標準化
X_scalered = scaler.transform(X)

# 学習済モデルで予測
predicted_label = gbm.predict(X_scalered)
predicted_label = predicted_label.reshape(-1,1)

# 提出できる形にIDとラベルの予測値（確率）を結合する
ID = np.array(df["SK_ID_CURR"]).reshape(-1,1)
submission3_values = np.concatenate([ID,predicted_label],axis=1)
submission3_values = pd.DataFrame(submission3_values,columns=["SK_ID_CURR","TARGET"])
submission3_values["SK_ID_CURR"] = submission3_values["SK_ID_CURR"].astype("int32")
print(submission3_values.head())

# csvに変換
submission3_values.to_csv('submission3.csv',index=False)

   SK_ID_CURR    TARGET
0      100001  0.036551
1      100005  0.069452
2      100013  0.033297
3      100028  0.047649
4      100038  0.128504


提出スコア：0.71046